src/librustc_typeck/collect.rs

```diff
@@ -1972,6 +2011,16 @@ fn explicit_predicates_of<'a, 'tcx>(
         );
     }
 
+    // Lint duplicate auto traits in bounds.
+    let auto_traits = predicates.iter().filter_map(|(pred, sp)| {
+        match pred {
+            ty::Predicate::Trait(ref pred) if tcx.trait_is_auto(pred.def_id()) =>
+                Some((Some(pred.skip_binder().self_ty()), pred.def_id(), *sp)),
```

Comment:
I don't love skipping the binder here. It's kind of surprising to get a `Ty` that has an implicit binder around it. You could use `pred.map_bound(|p| p.self_ty())` to get a `Binder<Ty<'tcx>>` instead.

Reply:
(I'm surprised there is no method on `PolyTraitRef` for that, but anyway.)

Reply:
A `self_ty` method on `PolyTraitRef`? Sure, I can add that, then use it.